components/core/src/clp/streaming_compression/zstd/Compressor.cpp

```diff
@@ -1,14 +1,51 @@
 #include "Compressor.hpp"
 
-#include "../../Defs.h"
-#include "../../spdlog_with_specializations.hpp"
+#include <cstddef>
+
+#include <spdlog/spdlog.h>
+#include <zstd_errors.h>
+
+#include "../../Array.hpp"
+#include "../../ErrorCode.hpp"
+#include "../../FileWriter.hpp"
+#include "../../TraceableException.hpp"
+#include "../Compressor.hpp"
+#include "../Constants.hpp"
+
+namespace {
+/**
+ * Checks if a value returned by ZStd function indicates an error code.
+ *
+ * For most ZStd functions that return `size_t` results, instead of returning a union type that can
+ * either be a valid result or an error code, an unanimous `size_t` type is returned.
+ * Usually, if the return value exceeds the maximum possible value of valid results, it is treated
+ * as an error code. However, the exact behavior is function-dependent, so ZStd provides:
+ * 1. A value checking function `ZSTD_isError`
+ * 2. A size_t <-> error_code_enum mapping function `ZSTD_getErrorCode`.
+ * See also: https://facebook.github.io/zstd/zstd_manual.html
+ *
+ * @param result A `size_t` type result returned by ZStd functions
+ * @return Whether the result is an error code and indicates an error has occurred
+ */
+auto is_error(size_t result) -> bool {
+    return 0 != ZSTD_isError(result) && ZSTD_error_no_error != ZSTD_getErrorCode(result);
+}
+}  // namespace
 
 namespace clp::streaming_compression::zstd {
 Compressor::Compressor()
-        : ::clp::streaming_compression::Compressor(CompressorType::ZSTD),
-          m_compression_stream_contains_data(false),
-          m_compressed_stream_file_writer(nullptr) {
-    m_compression_stream = ZSTD_createCStream();
+        : ::clp::streaming_compression::Compressor{CompressorType::ZSTD},
+          m_compressed_stream_file_writer{nullptr},
+          m_compression_stream{ZSTD_createCStream()},
+          m_compression_stream_contains_data{false},
+          m_compressed_stream_block_size{ZSTD_CStreamOutSize()},
+          m_compressed_stream_block_buffer{Array<char>{m_compressed_stream_block_size}},
+          m_compressed_stream_block{
+                  .dst = m_compressed_stream_block_buffer.data(),
+                  .size = m_compressed_stream_block_size,
+                  .pos = 0
+          },
+          m_uncompressed_stream_pos{0} {
```

Comment:
I think for most of initializations, including `m_compression_stream` that calls `ZSTD_createCSTream()`, can be moved into the member declaration. This is our latest guideline for default init: (https://www.notion.so/yscope/WIP-Coding-Guidelines-9a308b847a5343958ba3cb97a850be66?pvs=4#cd5d55358a6c4d629daa5673778f6436)